enhancements/storage/manila-csi-driver-operator.md

```diff
@@ -0,0 +1,221 @@
+
+To maintain the lifecycle of the driver we want to implement an operator, that will handle all administrative tasks: deploy, restore, upgrade, healthchecks, and so on.
+
+### Action plan
+
+#### Find a testing/development platform that supports Manila
+
+For testing and development we need a public OpenStack cloud with Manila support. So far MOC and PSI do not have this capability.
+The cloud should be OSP 13+ based, and comply with the reference architecture.
+There are two possibilities how we can accomplish this task:
+
+- Add Manila support to either MOC or PSI.
+
+- Find a completely new cloud, like next-gen RDO.
+
+Actions:
+
+- The first item on the list seems more likely. We need to contact the clouds admins and ask them to add Manila support.
+
+#### Update cloud-provider-openstack for OpenShift (Done)
+
+Currently the repo containing Manila CSI driver is forked, but seriously outdated https://github.com/openshift/cloud-provider-openstack
+We need to fetch all the latest changes from upstream.
+Actions:
+
+- Add ShiftStack team to the Owners (Done: https://github.com/openshift/cloud-provider-openstack/pull/15)
+
+- Prepare and merge a bump commit (Done: https://github.com/openshift/cloud-provider-openstack/pull/16)
+
+#### Build Manila CSI driver image and put it in Quay
+
+To start using Manila CSI in OpenShift we need to build an image and put it in the OCP account in Quay along with other OCP images.
+The driver provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in Quay in quay.io/k8scsi account.
+Actions:
+
+- Create a new repo for manila csi driver image in OCP quay account.
+
+- Set up auto image building based on the changes in the cloud provider repo.
+
+#### Test the driver manually
+
+When all required components are built, we can manually deploy the driver with NFS backend and test how it works.
+Actions:
+
+- Manually deploy the driver on a cloud with Manila support
+
+- Configure it to use NFS backend
+
+- Go through the lifecycle of volume (create, update, mount, write data, delete, etc.)
+
+#### Write Manila CSI driver operator
+
+The operator should be able to create, configure and manage Manila CSI driver for Kubernetes and OpenShift. In other words, automate what has been done on the previous step.
+Actions:
+
+- Create a new repo in OpenShift’s github: github.com/openshift/manila-csi-driver-operator
+
+- Implement the driver, using [OpenShift Operator SDK](https://docs.openshift.com/container-platform/4.1/applications/operator_sdk/osdk-getting-started.html)
+
+#### Add automatic testing
+
+To make sure that everything works properly, we need the support of CI. If Manila is added to MOC, we just need to run a few tests in e2e-openstack pipeline. K8s already implements all necessary tests for CSI drivers, and they should run automatically if the testing system detects a CSI driver attached.
```

Comment:
OCP has CI that runs all e2e tests on OpenStack. It can run a test that checks Manila CSI driver operator is installed when running on OpenStack. Or even run CSI e2e tests for it. However, I don't know which OpenStack the CI uses and if Manila is there.

Example of such run (with links to detailed logs, maybe you can find the OpenStack instance there): https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-openstack-4.3/574